### PR TITLE
Add KafkaConsumerHook and KafkaSensor

### PR DESCRIPTION
**Pull request by yyyyyyuanfei**
Dear Airflow Maintainers,

Please accept the following PR that
- Add the `KafkaConsumerHook`.
- Add the `KafkaSensor` which listens to messages with the specific topic.

Related Issue:
#1311

Reminder to contributors:
- You must add an Apache License header to all new files
- Please squash your commits when possible and follow the [7 rules of good Git commits](http://chris.beams.io/posts/git-commit/#seven-rules)

I am new to the community, I am not sure the files are at the right place or missing anything. 

The sensor could be used as the first node of a dag where the second node can be a `TriggerDagRunOperator`. The messages are polled in a batch and the dag runs are dynamically generated.  

Thanks!

### CODE REVIEW

**Comment by artwr** [comment]
@yuanfeiz Sorry it took us time to get to this. Overall, I think the idea is worthwhile, but I am not sure about the implementation. The hook looks fine, but I am worried about the sensor returning data. I do not think we have a good story for queue triggered DAGs just yet, but if you still want to go forward, happy to work with you on a design that makes sense.

**Comment by artwr** [comment]
@yuanfeiz Also, we are trying to clean up open PRs, so if I do not hear from you, I will close the PR.

**Comment by yyyyyyuanfei** [comment]
Hi @artwr, really glad to see updates on this thread. Basically we had been using it as a background job coordinator and executors with DAG. 

Sorry it has been a while I haven't kept up with the community, so I might be lack of the recent development. Is it the use case Airflow intends to cover? Let me know if there is anything I can help.

**Comment by artwr** [comment]
@yuanfeiz 
Background job coordinator and executor is definitely Airflow's use case. That said, we usually expect the heavy lifting to be done by the various systems Airflow talks to, not usually the airflow workers.

What this means is that usually, Airflow tasks are either : 
1) A sensor, which checks for a particular condition and returns True or False.
2) An operator which does delegates some work to a system (Say for instance a database, Spark or such)
3) A transfer operator, where we move data from one system to another, sometimes by using the local filesystem as an intermediate.

Airflow does not usually have tasks output data in the Python sense, mostly because there is no easy way in a distributed system to guarantee that the next task will be executed on the same worker, so loading the output of a task in memory is not usually useful.

For small amounts of data there the [XCom system](https://airflow.incubator.apache.org/concepts.html#xcoms) , which aims at being a lightweight cross communication system between tasks.

Maybe you could describe your use case a little better in a JIRA task [here](https://issues.apache.org/jira/browse/AIRFLOW) or on the dev mailing list dev@airflow.incubator.apache.org, and we could see where to take this.

Let me know if you have any more questions.

**Comment by r39132** [comment]
Closing.. not likely to merge.

**Comment by chanansh** [comment]
@r39132 we use Kafka to send a message "run now" as trigger.
Can you please merge this solution?

**Comment by ashb** [comment]
@chanansh If you want this you could just include these two hook and operators in your local install. If you think it makes sense to include in stock Airflow you can open a new PR, but please ensure the include in the docs that this is not suitable for low-latency/high-throughput/streaming.

**Comment by dferguson992** [comment]
@ashb can we consider re-opening this PR?  I'd be happy to update the docs for @chanansh but am hesitant to submit a full PR for work i didn't write originally

**Comment by ashb** [comment]
@dferguson992 You can "adopt" it by putting the correct `Co-authored-by` line at the foot of the commit. https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors

It's so old that re-opening now is not possible (github won't let me)

**Comment by dferguson992** [comment]
I'll give @chanansh a few days to re-open and am happy to hop on as co-author as per your linked doc to fix the documentation issues in your [previous comment](https://github.com/apache/airflow/pull/1415#issuecomment-484429806).  If not I'll just re-open under my name.  It's good work and I think Kafka and Airflow is a great integration, especially concerning data quality management.

**Comment by dferguson992** [comment]
@ashb I've created #7407 with the co-authoring steps, but I cannot find the original AIRFLOW-XXXX issue referenced in this original commit.  Thoughts on how to proceed?

File: airflow/contrib/hooks/kafka_hook.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        self.server = '{host}:{port}'.format(**locals())
+        self.consumer = KafkaConsumer(
+            self.topic,
+            bootstrap_servers=self.server, **conf)
+
+        return self.consumer
+
+    def get_messages(self):
+        """
+        Get all the messages haven't been consumed, it doesn't
```
Comment (luup2k) [inline airflow/contrib/hooks/kafka_hook.py:47]: > "Get all the messages haven't been consumed,"

If we use poll() without max_records, the behavior is returns at most "max_poll_records" #records. "max_poll_records" is setted to 500 by default at Consumer Init config.
So, we're not going to consume "all" message except we put a very high number as max_poll_records(could be a memory bomb) or  we have a low number of message in the topic.

https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html#kafka.KafkaConsumer.poll

**Pull request closed without merging**